Add update facility location dashboard page

jwalgran opened the pull request:
## Overview

Adds a new API and dashboard page to update the coordinates of a `Facility`.

Connects #738

## Demo

![2019-09-18 19 50 09](https://user-images.githubusercontent.com/17363/65209728-c0977880-da4d-11e9-91e5-b1ac08edbd46.gif)

## Notes

Changes to the facility location made by superusers will be recorded as records in the new `FacilityLocation` table, allowing us to capture details about the moderation activity.

We exclude the new endpoint from the Swagger docs because it is intended to be consumed by the superuser-only dashboard.

The layout of the new dashboard page is based on other "two panel" pages, like facility merging.

We added "Location" to the `DashboardFacilityCard` so the existing coordinates are visible. This adds the location to every dashboard page that uses the `DashboardFacilityCard` but showing extra information is not a problem.

## Testing Instructions

* Search the facilities and copy an OAR ID.
* Browse http://localhost:6543/dashboard/updatefacilitylocation.
* Enter the OAR ID and click search. Verify that the facility details are loaded.
* Copy the longitude and latitude from the facility details to the "Longitude" and "Latitude" fields and verify that a map showing the same location is loaded. Ensure that there are only 5 decimal places.
* Adjust the coordinates to move the point.
* Click "Update Location." Verify that the map updates, the New Location fields are cleared, and a success toast message is shown.
* Repeat the process, this time filling in the "Contributor" and "Notes" fields. Click "Update Location."
* Run `./scripts/dbshell` and run `SELECT * FROM api_facilitylocation;`. Verify that the table contains rows for each of the saved location changes.

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

rajadain commented:
Taking a look now.

rajadain commented on src/app/src/components/DashboardUpdateFacilityLocationCard.jsx:
```diff
@@ -0,0 +1,246 @@
+                    />
+                    <TextField
+                        variant="outlined"
+                        placeholder="Latitue"
```
Typo `Latitue` -> `Latitude`

jwalgran replied:
Fixed in 9fd3fd7

rajadain commented:
I'm seeing more than 5 decimal places:

![image](https://user-images.githubusercontent.com/1430060/65262268-f3864e80-dad7-11e9-9fa8-2f0dd94d66e1.png)

Also, should we allow them to update the location when it is the exact same location?

rajadain commented:
The order of coordinates between "Old Location" and "New Location" is flipped:

![image](https://user-images.githubusercontent.com/1430060/65262393-3d6f3480-dad8-11e9-9d19-966a4e2cf48e.png)

rajadain commented:
When I tried to submit the above results I got an error from the server admonishing the presence of more than 5 decimal places, but this was not conveyed to the user anywhere:

![image](https://user-images.githubusercontent.com/1430060/65262489-73acb400-dad8-11e9-8d9c-c4c7b63bae83.png)

rajadain commented:
Seeing the correct values in the database:

```sql
SELECT id, ST_AsText(location), notes, created_at, updated_at, contributor_id, created_by_id, facility_id 
FROM api_facilitylocation;

 id |       st_astext       |       notes       |          created_at           |          updated_at           | contributor_id | created_by_id |   facility_id
----+-----------------------+-------------------+-------------------------------+-------------------------------+----------------+---------------+-----------------
  1 | POINT(75.2304 23.039) |                   | 2019-09-19 16:25:28.876645+00 | 2019-09-19 16:25:28.876694+00 |                |             1 | IN20192484GK1SJ
  2 | POINT(75.2304 23.139) | Test Change Notes | 2019-09-19 16:29:45.830376+00 | 2019-09-19 16:29:45.830399+00 |              8 |             1 | IN20192484GK1SJ

(2 rows)
```

jwalgran commented:
I have pushed up commits to address all the comments.

>I'm seeing more than 5 decimal places:

Google returns 7, which is the "practical limit of commercial surveying" (https://en.wikipedia.org/wiki/Decimal_degrees) I increased the decimal length to 7 in fixup 03b80ed

>Also, should we allow them to update the location when it is the exact same location?

Not reall harmful, but not helpful and potentially confusing. Disallowed in fixup 6eefa19

>The order of coordinates between "Old Location" and "New Location" is flipped

Fixed in 72b6509

>When I tried to submit the above results I got an error [...] but this was not conveyed to the user anywhere:

Added quick and dirty error display in fixup da72a4d (which should be fine for this admin-only page)

rajadain commented:
There seem to be some linting errors holding up the build

jwalgran commented:
Lint fixed

rajadain commented:
This may be out of scope of this card, but the order of coordinates in the detail sidebar is opposite that of this screen:

![image](https://user-images.githubusercontent.com/1430060/65267555-40235700-dae3-11e9-904e-4bd98361a783.png)

Also, I don't see the "Other Locations" even though I just added some:

```sql
SELECT id, ST_AsText(location), notes, created_at, updated_at, contributor_id, created_by_id, facility_id 
FROM api_facilitylocation
WHERE facility_id = 'US20192484EJ6T6';

 id |          st_astext           |        notes        |          created_at           |          updated_at           | contributor_id | created_by_id |   facility_id
----+------------------------------+---------------------+-------------------------------+-------------------------------+----------------+---------------+-----------------
  3 | POINT(-81.3044164 28.7313)   |                     | 2019-09-19 17:39:37.704818+00 | 2019-09-19 17:39:37.704844+00 |                |             1 | US20192484EJ6T6
  4 | POINT(-81.3044164 28.731334) | Restore to original | 2019-09-19 17:40:16.467718+00 | 2019-09-19 17:40:16.467746+00 |              7 |             1 | US20192484EJ6T6

(2 rows)
```

rajadain commented:
Ah I see that's been extracted to a future issue.

jwalgran commented:
Yes. Extracted into #823. There are some additional details that Katie requested after seeing the draft feature on our weekly check in call.

jwalgran commented:
> This may be out of scope of this card, but the order of coordinates in the detail sidebar is opposite that of this screen

I will add that to #823, which will cover multiple changes to location display in the sidebar.

rajadain commented on src/app/src/actions/updateFacilityLocation.js:
```diff
@@ -0,0 +1,47 @@
+    createAction('COMPLETE_FETCH_UPDATE_LOCATION_FACILITY');
+export const clearUpdateLocationFacility = createAction('CLEAR_UPDATE_LOCATION_FACILITY');
+export const updateUpdateLocationFacilityOARID =
+    createAction('UPDATE_UPDATE_LOCATION_FACILITY_OAR_ID');
```
Unfortunate naming here, not sure how it can be helped.

jwalgran commented on src/django/api/views.py:
```diff
@@ -1154,6 +1161,45 @@ def promote(self, request, pk=None):
+                Contributor.objects.get(id=contributor_id)
+        facility_location.save()
+
+        facility.location = facility_location.location
```
Yes. The `HistoricalFacility` managed by `django-simple-history` will have the previous location value.

jwalgran commented:
Thanks for refining this with me.
